cpp/src/arrow/compute/kernels/codegen_internal.cc

```diff
@@ -56,9 +56,16 @@ Result<TypeHolder> LastType(KernelContext*, const std::vector<TypeHolder>& types
   return types.back();
 }
 
-Result<TypeHolder> ListValuesType(KernelContext*, const std::vector<TypeHolder>& args) {
-  const auto& list_type = checked_cast<const BaseListType&>(*args[0].type);
-  return list_type.value_type().get();
+Result<TypeHolder> ListValuesType(KernelContext* ctx,
+                                  const std::vector<TypeHolder>& args) {
+  auto list_type = checked_cast<const BaseListType*>(args[0].type);
+  auto value_type = list_type->value_type().get();
+  for (auto value_kind = value_type->id();
+       is_list(value_kind) || is_list_view(value_kind); value_kind = value_type->id()) {
+    list_type = checked_cast<const BaseListType*>(list_type->value_type().get());
+    value_type = list_type->value_type().get();
+  }
```

Comment:
Previous implementation of this function would return a `list<int32>` type if you passed `list<list<int32>>`. Not it returns `int32`. Are you sure that is safe for all the usages? If yes, the name should reflect that it's recursive now. Since "list values type" has a well-defined meaning in the Arrow types.

Reply:
Ah, i understand what you mean. It's not correct!